Unable to unwrap `Option` struct field in `if var == none {` expr

### Describe the bug

V automatically unwraps the `Option` in expressions like this:

```v
mut a := ?int(none)
a = 1
if a == none {
  // a is Option(none)
} else {
  // a is unwrapped to int: 1
}
```

See the example: https://play.vlang.io/p/423d702cd1

<details><summary>Example code...</summary>
<p>

```v
fn unwrap(a ?[3]u8) {
	if a == none {
		println('${@FN}:${@LINE}: ${typeof(a).name}')
	} else {
		println('${@FN}:${@LINE}: ${typeof(a).name}')
	}
}

struct Opt {
pub mut:
	f ?[3]u8
}

fn unwrap_field(opt Opt) {
	if opt.f == none {
		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
	} else {
		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
	}
}

fn main() {
	mut a := ?[3]u8(none)
	unwrap(a)
	a = [3]u8{init: 1}
	unwrap(a)

	mut f := Opt{}
	unwrap_field(f)
	f.f = [3]u8{init: 2}
	unwrap_field(f)
}
```

</p>
</details>

Note the output:
```
unwrap:3: ?[3]u8
unwrap:5: [3]u8
unwrap_field:16: ?[3]u8
unwrap_field:18: ?[3]u8
```

`a` is unwrapped in `if` expression and optional struct field `Opt.f` not.

### Reproduction Steps

Try to unwrap field manually (https://play.vlang.io/p/06491b51fe):

<details><summary>Example code...</summary>
<p>

```v
struct Opt {
pub mut:
	f ?[3]u8
}

fn unwrap_field(opt Opt) {
	if opt.f == none {
		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
	} else {
		x := opt.f as [3]u8
		dump(x)
		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
	}
}

fn main() {
	mut f := Opt{}
	unwrap_field(f)
	f.f = [3]u8{init: 2}
	unwrap_field(f)
}
```

</p>
</details>

The result is C error:

```console
$ v -cg -show-c-output run /tmp/unwrap.vv
======== Output of the C Compiler (/home/ge/.vlang/thirdparty/tcc/tcc.exe) ========
/tmp/v_1000/unwrap.01JQSDXBZ73T0PP7MM0509TV1E.tmp.c:6065: error: '{' expected (got ";")
===================================================================================
======== Output of the C Compiler (cc) ========
/tmp/v_1000/unwrap.01JQSDXBZ73T0PP7MM0509TV1E.tmp.c: In function ‘main__unwrap_field’:
/tmp/v_1000/unwrap.01JQSDXBZ73T0PP7MM0509TV1E.tmp.c:6065:38: error: invalid initializer
 6065 |                 Array_fixed_u8_3 x = (*(Array_fixed_u8_3*)opt.f.data);
      |                                      ^
===============================================
No such file or directory; code: 2
```

### Expected Behavior

Optional field is automatically unwrapped in `if` expression.

### Current Behavior

Option field is not unwrapped. Invalid C code generates on manual unwrapping.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.10 42ac688

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.10 89d1aac.42ac688
|:-------------------|:-------------------
|OS                  |linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
|Processor           |16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P
|Memory              |2.56GB/15.35GB
|                    |
|V executable        |/home/ge/.vlang/v
|V last modified time|2025-04-01 19:43:21
|                    |
|V home dir          |OK, value: /home/ge/.vlang
|VMODULES            |OK, value: /home/ge/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/ge
|                    |
|Git version         |git version 2.49.0
|V git status        |weekly.2025.13-42-g42ac6885
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20250207
|gcc version         |gcc (GCC) 14.2.1 20250207
|clang version       |clang version 19.1.7
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.41

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/if.v
+++ b/vlib/v/checker/if.v
@@ -691,7 +691,8 @@ fn (mut c Checker) smartcast_if_conds(mut node ast.Expr, mut scope ast.Scope, co
 		mut first_cond := control_expr.branches[0].cond
 		// handles unwrapping on if var == none { /**/ } else { /*unwrapped var*/ }
 		if mut first_cond is ast.InfixExpr {
-			if first_cond.left is ast.Ident && first_cond.op == .eq && first_cond.right is ast.None {
+			if first_cond.left in [ast.Ident, ast.SelectorExpr] && first_cond.op == .eq
+				&& first_cond.right is ast.None {
 				if c.comptime.get_ct_type_var(first_cond.left) == .smartcast {
 					first_cond.left_type = c.type_resolver.get_type(first_cond.left)
 					c.smartcast(mut first_cond.left, first_cond.left_type, first_cond.left_type.clear_flag(.option), mut

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4185,7 +4185,8 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 					for i, typ in field.smartcasts {
 						if i == 0 && (is_option_unwrap || nested_unwrap) {
 							deref := if g.inside_selector {
-								if is_iface_or_sumtype {
+								if is_iface_or_sumtype || (field.orig_type.is_ptr() && g.left_is_opt
+									&& is_option_unwrap) {
 									'*'.repeat(field.smartcasts.last().nr_muls())
 								} else {
 									'*'.repeat(field.smartcasts.last().nr_muls() + 1)

--- a/vlib/v/tests/loops/for_smartcast_test.v
+++ b/vlib/v/tests/loops/for_smartcast_test.v
@@ -16,7 +16,7 @@ fn test_nested_sumtype_selector() {
 		pos: 1
 	}))}
 	for c.node is Expr {
-		assert typeof(c.node).name == 'Node'
+		assert typeof(c.node).name == 'Expr'
 		break
 	}
 }

--- a/vlib/v/tests/options/option_if_unwrap_test.v
+++ b/vlib/v/tests/options/option_if_unwrap_test.v
@@ -1,0 +1,54 @@
+fn unwrap(a ?[3]u8) {
+	if a == none {
+		println('${@FN}:${@LINE}: ${typeof(a).name}')
+		assert typeof(a).name == '?[3]u8'
+	} else {
+		println('${@FN}:${@LINE}: ${typeof(a).name}')
+		assert typeof(a).name == '[3]u8'
+	}
+	if a != none {
+		println('${@FN}:${@LINE}: ${typeof(a).name}')
+		assert typeof(a).name == '[3]u8'
+	} else {
+		println('${@FN}:${@LINE}: ${typeof(a).name}')
+		assert typeof(a).name == '?[3]u8'
+	}
+}
+
+struct Opt {
+pub mut:
+	f ?[3]u8
+}
+
+fn unwrap_field(opt Opt) {
+	if opt.f == none {
+		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
+		assert opt.f == ?[3]u8(none)
+		assert typeof(opt.f).name == '?[3]u8'
+	} else {
+		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
+		assert opt.f == [u8(2), 2, 2]!
+		assert typeof(opt.f).name == '[3]u8'
+	}
+	if opt.f != none {
+		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
+		assert opt.f == [u8(2), 2, 2]!
+		assert typeof(opt.f).name == '[3]u8'
+	} else {
+		println('${@FN}:${@LINE}: ${typeof(opt.f).name}')
+		assert opt.f == ?[3]u8(none)
+		assert typeof(opt.f).name == '?[3]u8'
+	}
+}
+
+fn test_main() {
+	mut a := ?[3]u8(none)
+	unwrap(a)
+	a = [3]u8{init: 1}
+	unwrap(a)
+
+	mut f := Opt{}
+	unwrap_field(f)
+	f.f = [3]u8{init: 2}
+	unwrap_field(f)
+}

--- a/vlib/v/tests/options/option_ptr_unwrap_test.v
+++ b/vlib/v/tests/options/option_ptr_unwrap_test.v
@@ -27,9 +27,9 @@ fn (mut li LinkedList[T]) add[T](data T) ? {
 		node.prev = node
 	} else {
 		node.next = li.head
-		node.prev = li.head?.prev
+		node.prev = li.head.prev
 		node.prev?.next = node
-		li.head?.prev = node
+		li.head.prev = node
 	}
 
 	li.size += 1

--- a/vlib/v/tests/options/option_struct_init_with_ref_opt_test.v
+++ b/vlib/v/tests/options/option_struct_init_with_ref_opt_test.v
@@ -31,7 +31,7 @@ pub fn (mut l LinkedList) pop() int {
 		l.head = none
 		l.tail = none
 	} else { // multiple elements
-		mut prev := tail.prev or { panic('prev cannot be none at this point') }
+		mut prev := tail.prev
 		prev.next = none
 		l.tail = prev
 	}

--- a/vlib/v/type_resolver/comptime_resolver.v
+++ b/vlib/v/type_resolver/comptime_resolver.v
@@ -75,6 +75,11 @@ pub fn (mut t TypeResolver) typeof_type(node ast.Expr, default_type ast.Type) as
 		if node.expr is ast.Ident && node.is_field_typ {
 			return t.get_type_from_comptime_var(node.expr)
 		}
+		if field := node.scope.find_struct_field(node.expr.str(), node.expr_type, node.field_name) {
+			if field.smartcasts.len > 0 {
+				return field.smartcasts.last()
+			}
+		}
 		sym := t.table.sym(t.resolver.unwrap_generic(node.expr_type))
 		if f := t.table.find_field_with_embeds(sym, node.field_name) {
 			return f.typ

